### PR TITLE
compiler: optimize map[string([]byte))] lookups

### PR DESCRIPTION
**Pull request by dgryski**

### CODE REVIEW

**Comment by dgryski** [comment]
This doesn't seem to work (according to the benchmarks in https://github.com/tinygo-org/tinygo/issues/2699 ) and I have no idea why.

**Comment by dgryski** [comment]
So, my PR adds is the definitions for `%5` and `%6`, but `%3`, `%4`, and the call to `runtme.trackPointer` still seem to be hanging around.  Not sure how to remove those.  Probably at a later stage in a transform optimization pass.

```
  %3 = call %runtime._string @runtime.stringFromBytes(i8* %b.data, i32 %b.len, i32 %b.cap, i8* undef), !dbg !62164
  %4 = extractvalue %runtime._string %3, 0, !dbg !62164
  call void @runtime.trackPointer(i8* %4, i8* undef), !dbg !62164
  %5 = insertvalue { i8*, i32 } undef, i8* %b.data, 0, !dbg !62165
  %6 = insertvalue { i8*, i32 } %5, i32 %b.len, 1, !dbg !62165
```

**Comment by FrankReh** [comment]
okay. glad you are making progress in understanding this. I was wondering why I didn't see it working either. I really don't understand that level of IR or the passes the tinygo compiler is making so not sure how I can help at the moment.

**Comment by aykevl** [comment]
There is a very similar optimization already: `OptimizeStringToBytes` (see transform/rtcalls.go). But you want the opposite, essentially.

I'm thinking of ways to make this `[]byte`-to-`string` optimization more general. But it's kind of difficult. For the `string`-to-`[]byte` conversion optimization, it's easy: just check that none of the uses of the conversion modify the slice. But the other way round is hard because even though that particular function might not modify the slice, there can be aliases that might be modified.

File: compiler/compiler.go [hunk 1]
```diff
@@ -1775,6 +1776,40 @@ func (b *builder) createExpr(expr ssa.Value) (llvm.Value, error) {
...
+								key := b.getValue(convert.X)
+
+								keyBuf := b.CreateExtractValue(key, 0, "map.keyBuf")
+								keyLen := b.CreateExtractValue(key, 1, "map.keyLen")
+
+								// Create the string
+								str := b.ctx.ConstStruct([]llvm.Value{
+									llvm.Undef(b.i8ptrType),
+									llvm.Undef(b.uintptrType),
+								}, false)
```
Comment (aykevl) [inline compiler/compiler.go:1798]: This looks unsafe. A string is not just a struct, it's a `runtime._string` named struct. ~And I don't think these values you're inserting are actually constants.~ nevermind misread the code.

**Comment by dgryski** [comment]
This is not the approach.

**Pull request closed without merging**